### ISSUE TITLE
Add walk() method to traverse Arr, Hash and other containers in uniform way

### ISSUE THREAD

**Issue by ilyash-b** [ngs-lang/ngs#419]

**Comment by ilyash** [comment]
Requirements:
* breadth-first / depth-first
* support transformation (map-like behavior)
* support filtering (filter-like behavior)
* provide context to the callback (path, etc), not only current value
* support passing arbitrary values from the callback to other callbacks invocations

(will probably think of additional points later)

**Comment by ilyash-b** [comment]
Use case to be covered: find a Hash with `{'linkId': Str}` at any level

**Comment by ilyash-b** [comment]
To consider: prune functionality? (callback will be called additional "controller" parameter which will have `prune()` method)

**Comment by ilyash-b** [comment]
Minimalistic implementation from a project:

```
	F walk(x, cb:Fun) {
		cb(x)
	}
	F walk(a:Arr, cb:Fun) {
		cb(a)
		a.each(walk(X, cb))
	}
	F walk(h:Hash, cb:Fun) {
		cb(h)
		h.eachv(walk(X, cb))
	}
```

**Comment by ilyash** [comment]
Merged with #437
